go-controller/pkg/ovn/egressfirewall_dns.go

```diff
@@ -0,0 +1,195 @@
+		dnsEntries:        make(map[string]*dnsEntry),
+		addressSetFactory: addressSetFactory,
+
+		added:          make(chan bool),
+		stopChan:       make(chan struct{}),
+		controllerStop: controllerStop,
+	}
+
+	go utilwait.Until(egressDNS.Sync, 0, egressDNS.stopChan)
+
+	return egressDNS, nil
+}
+
+func (e *EgressDNS) Add(namespace, dnsName string) (AddressSet, error) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	if _, exists := e.dnsEntries[dnsName]; !exists {
+		var err error
+		dnsEntry := dnsEntry{
+			namespaces: make(map[string]struct{}),
+		}
+		if e.addressSetFactory == nil {
+			panic("addressSetFactory is null, that cannot happen")
+		}
+		dnsEntry.dnsAddressSet, err = e.addressSetFactory.NewAddressSet(dnsName, nil)
+		if err != nil {
+			return nil, fmt.Errorf("cannot create addressSet for %s: %v", dnsName, err)
+		}
+		e.dnsEntries[dnsName] = &dnsEntry
+		//only call Add if the dnsName doesn't exist in dnsEntries
+		if err := e.dns.Add(dnsName); err != nil {
+			klog.Errorf("DNS Name: %s cannot be added to the dnsMap - %v", dnsName, err)
+		}
+		unsorted := e.dns.GetIPs(dnsName)
+		e.dnsEntries[dnsName].dnsResolves = SortIPs(unsorted)
+		err = e.dnsEntries[dnsName].dnsAddressSet.AddIPs(e.dnsEntries[dnsName].dnsResolves)
+		if err != nil {
+			return nil, fmt.Errorf("cannot add IPs to EgressFirewalls AddressSet: %s - %v", dnsName, err)
+		}
+		e.signalAdded()
+	}
+	e.dnsEntries[dnsName].namespaces[namespace] = struct{}{}
+	return e.dnsEntries[dnsName].dnsAddressSet, nil
+
+}
+
+func (e *EgressDNS) Delete(namespace string) bool {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	// go through all dnsNames for namespaces
+	for dnsName, dnsEntry := range e.dnsEntries {
+		// delete the dnsEntry
+		delete(dnsEntry.namespaces, namespace)
+		if len(dnsEntry.namespaces) == 0 {
+			// the dnsEntry appears in no other namespace so delete the address_set
+			err := dnsEntry.dnsAddressSet.Destroy()
+			if err != nil {
+				klog.Errorf("Error deleteing EgressFirewall AddressSet for dnsName: %s %v", dnsName, err)
+			}
+			// the dnsEntry is no longer needed because nothing referances it delete it
+			delete(e.dnsEntries, dnsName)
+		}
+	}
+	return len(e.dnsEntries) == 0
+}
+
+func (e *EgressDNS) Update(dns string) (bool, error) {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	return e.dns.Update(dns)
+}
+
+func (e *EgressDNS) Sync() {
+	var duration time.Duration
+	for {
+		tm, dnsName, ok := e.GetNextQueryTime()
+		if !ok {
+			duration = 30 * time.Minute
+		} else {
+			now := time.Now()
+			if tm.After(now) {
+				duration = tm.Sub(now)
+			} else {
+				_, err := e.Update(dnsName)
+				if err != nil {
+					utilruntime.HandleError(err)
+				}
+
+				sortedIPs := SortIPs(e.dns.GetIPs(dnsName))
+				if !reflect.DeepEqual(e.dnsEntries[dnsName].dnsResolves, sortedIPs) {
+					//flush the old ips from the addressSets
+					addressSet := e.dnsEntries[dnsName].dnsAddressSet
+					err := addressSet.DeleteIPs(e.dnsEntries[dnsName].dnsResolves)
```

Comment:
Don't flush and re-add, compute the delta.

Reply:
While you're at it, can you make `address_set.go` batched? This seems really inefficient.